promote error

## Summary 

What happened/what you expected to happen?
[root@node01 ~]# kubectl argo rollouts promote reviews-v1 -n istio
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17bfbfd]

goroutine 1 [running]:
github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote.getPatches(0xc000b79200, 0x2000000, 0xc0000520c8, 0x7ffeaa3e9714, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote/promote.go:168 +0x37d
github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote.PromoteRollout(0x26ca978, 0xc00034b160, 0x7ffeaa3e9714, 0xa, 0xc000000000, 0xc000196400, 0xc00091fd38, 0x1c8071f)
        /go/src/github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote/promote.go:113 +0x11d
github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote.NewCmdPromote.func1(0xc000198840, 0xc000163260, 0x1, 0x3, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/pkg/kubectl-argo-rollouts/cmd/promote/promote.go:68 +0x13b
github.com/spf13/cobra.(*Command).execute(0xc000198840, 0xc000163230, 0x3, 0x3, 0xc000198840, 0xc000163230)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc00013e840, 0xc00013e840, 0x2679840, 0xc00000e020)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
        /go/src/github.com/argoproj/argo-rollouts/cmd/kubectl-argo-rollouts/main.go:21 +0xc5
## Diagnostics

What version of Argo Rollouts are you running?
1.30
```
# Paste the logs from the rollout controller
istio
# Logs for the entire controller:
kubectl logs -n argo-rollouts deployment/argo-rollouts

# Logs for a specific rollout:
kubectl logs -n argo-rollouts deployment/argo-rollouts | grep rollout=<ROLLOUTNAME>
```

---
<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

--- a/pkg/kubectl-argo-rollouts/cmd/promote/promote.go
+++ b/pkg/kubectl-argo-rollouts/cmd/promote/promote.go
@@ -138,11 +138,14 @@ func getPatches(rollout *v1alpha1.Rollout, skipCurrentStep, skipAllStep, full bo
 		_, index := replicasetutil.GetCurrentCanaryStep(rollout)
 		// At this point, the controller knows that the rollout is a canary with steps and GetCurrentCanaryStep returns 0 if
 		// the index is not set in the rollout
-		if *index < int32(len(rollout.Spec.Strategy.Canary.Steps)) {
-			*index++
+		if index != nil {
+			if *index < int32(len(rollout.Spec.Strategy.Canary.Steps)) {
+				*index++
+			}
+			statusPatch = []byte(fmt.Sprintf(setCurrentStepIndex, *index))
+			unifiedPatch = statusPatch
 		}
-		statusPatch = []byte(fmt.Sprintf(setCurrentStepIndex, *index))
-		unifiedPatch = statusPatch
+
 	case skipAllStep:
 		statusPatch = []byte(fmt.Sprintf(setCurrentStepIndex, len(rollout.Spec.Strategy.Canary.Steps)))
 		unifiedPatch = statusPatch
@@ -165,11 +168,13 @@ func getPatches(rollout *v1alpha1.Rollout, skipCurrentStep, skipAllStep, full bo
 			_, index := replicasetutil.GetCurrentCanaryStep(rollout)
 			// At this point, the controller knows that the rollout is a canary with steps and GetCurrentCanaryStep returns 0 if
 			// the index is not set in the rollout
-			if *index < int32(len(rollout.Spec.Strategy.Canary.Steps)) {
-				*index++
+			if index != nil {
+				if *index < int32(len(rollout.Spec.Strategy.Canary.Steps)) {
+					*index++
+				}
+				statusPatch = []byte(fmt.Sprintf(clearPauseConditionsPatchWithStep, *index))
+				unifiedPatch = []byte(fmt.Sprintf(unpauseAndClearPauseConditionsPatchWithStep, *index))
 			}
-			statusPatch = []byte(fmt.Sprintf(clearPauseConditionsPatchWithStep, *index))
-			unifiedPatch = []byte(fmt.Sprintf(unpauseAndClearPauseConditionsPatchWithStep, *index))
 		}
 	}
 	return specPatch, statusPatch, unifiedPatch

--- a/pkg/kubectl-argo-rollouts/cmd/promote/promote_test.go
+++ b/pkg/kubectl-argo-rollouts/cmd/promote/promote_test.go
@@ -94,6 +94,31 @@ func TestPromoteSkipFlagOnNoStepCanaryError(t *testing.T) {
 	assert.Contains(t, stderr, skipFlagWithNoStepCanaryError)
 }
 
+func TestPromoteNoStepCanary(t *testing.T) {
+	ro := v1alpha1.Rollout{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "guestbook",
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.RolloutSpec{
+			Strategy: v1alpha1.RolloutStrategy{
+				Canary: &v1alpha1.CanaryStrategy{},
+			},
+		},
+	}
+	tf, o := options.NewFakeArgoRolloutsOptions(&ro)
+	defer tf.Cleanup()
+	cmd := NewCmdPromote(o)
+	cmd.PersistentPreRunE = o.PersistentPreRunE
+	cmd.SetArgs([]string{"guestbook"})
+	err := cmd.Execute()
+	assert.NoError(t, err)
+	stdout := o.Out.(*bytes.Buffer).String()
+	assert.NotEmpty(t, stdout)
+	stderr := o.ErrOut.(*bytes.Buffer).String()
+	assert.Empty(t, stderr)
+}
+
 func TestPromoteCmdSuccesSkipAllSteps(t *testing.T) {
 	ro := v1alpha1.Rollout{
 		ObjectMeta: metav1.ObjectMeta{
@@ -144,7 +169,7 @@ func TestPromoteCmdSuccesSkipAllSteps(t *testing.T) {
 	assert.Empty(t, stderr)
 }
 
-func TestPromoteCmdSuccesFirstStep(t *testing.T) {
+func TestPromoteCmdSuccesFirstStepWithSkipFirstStep(t *testing.T) {
 	ro := v1alpha1.Rollout{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "guestbook",
@@ -184,6 +209,56 @@ func TestPromoteCmdSuccesFirstStep(t *testing.T) {
 	cmd := NewCmdPromote(o)
 	cmd.PersistentPreRunE = o.PersistentPreRunE
 	cmd.SetArgs([]string{"guestbook", "-c"})
+
+	err := cmd.Execute()
+	assert.Nil(t, err)
+	assert.Equal(t, int32(1), *ro.Status.CurrentStepIndex)
+	stdout := o.Out.(*bytes.Buffer).String()
+	stderr := o.ErrOut.(*bytes.Buffer).String()
+	assert.Equal(t, stdout, "rollout 'guestbook' promoted\n")
+	assert.Empty(t, stderr)
+}
+
+func TestPromoteCmdSuccesFirstStep(t *testing.T) {
+	ro := v1alpha1.Rollout{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "guestbook",
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.RolloutSpec{
+			Strategy: v1alpha1.RolloutStrategy{
+				Canary: &v1alpha1.CanaryStrategy{
+					Steps: []v1alpha1.CanaryStep{
+						{
+							SetWeight: pointer.Int32Ptr(1),
+						},
+						{
+							SetWeight: pointer.Int32Ptr(2),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	tf, o := options.NewFakeArgoRolloutsOptions(&ro)
+	defer tf.Cleanup()
+	fakeClient := o.RolloutsClient.(*fakeroclient.Clientset)
+	fakeClient.PrependReactor("patch", "*", func(action kubetesting.Action) (handled bool, ret runtime.Object, err error) {
+		if patchAction, ok := action.(kubetesting.PatchAction); ok {
+			patchRo := v1alpha1.Rollout{}
+			err := json.Unmarshal(patchAction.GetPatch(), &patchRo)
+			if err != nil {
+				panic(err)
+			}
+			ro.Status.CurrentStepIndex = patchRo.Status.CurrentStepIndex
+		}
+		return true, &ro, nil
+	})
+
+	cmd := NewCmdPromote(o)
+	cmd.PersistentPreRunE = o.PersistentPreRunE
+	cmd.SetArgs([]string{"guestbook"})
 
 	err := cmd.Execute()
 	assert.Nil(t, err)

